### ISSUE TITLE
Continue performance tuning to squeeze out more performance.

### ISSUE THREAD

**Issue by ichard26** [ichard26/black-mypyc-wheels#2]
While lots of compatibility work has been done in https://github.com/psf/black/tree/mypyc-support-pt2, limited performance tuning has been done. It's very possible more performance can be squeezed out. Currently only a general (not even mypyc specific) strings optimization is present in that branch.

**Comment by ichard26** [comment]
Alright, with blackbench 21.7.dev2 finished, I've spent some time optimizing blib2to3 for mypyc. I was able to achieve a 16% performance bump on top of the existing 71-74% mypyc was able to do only with the compatibility changes. This was done via two optimization passes. Various optimizations implemented include:

- Adding `Final` so mypyc can use early binding (avoids expensive dictionary lookup at runtime) for multiple module-level and function-level variables;
- Using slightly faster code in \*very\* hot code (turns out `t, var_that_is_unused = tup` is more expensive than `t = tup[0]` under mypyc ... although not with the interpreter)
- Hard-coding function calls to allow even more early binding
- Reordering if-checks for the happy / common case so less checks occur overall
- Analyzing the logic and types for any redundant work that can be avoided (ie. removing useless asserts) 
- Tightening up declared types (so mypyc can generate better code)

At this point, I haven't been able to find any further optimization possibilities to check. For context, here's the hottest blib2to3 functions:

- `blib2to3.pgen2.parse.Parser.addtoken`
- `blib2to3.pgen2.parse.Parser.push`
- `blib2to3.pgen2.parse.Parser.pop`
- `blib2to3.pgen2.parse.Parser.shift`
- `blib2to3.pgen2.tokenize.generate_tokens`
- `blib2to3.pytree.convert`

I'll continue for a little bit longer but I think it's soon time to try my hand at optimizing the code under `src/black/` :)

**Comment by ichard26** [comment]
At this point I'm done optimizing /src/blib2to3 and am now trying to optimize /src/black. Key word being "trying" because unfortunately this codebase has a few challenges:

- There's a lot more code in /src/black which means there's more functions, classes, and other logic ...
- Which if not already more overwhelming is worsen by the fact the majority of the time is spread quite evenly across SO MANY FUNCTIONS. The lack of small contained hot spots as like with /src/blib2to3 really makes it a lot harder to figure out what and how to optimize. 

Now it's not a total loss, there's still a fair bit of stuff that looks promising (or at least like a potential optimizable chunk of code):

- `black.lines.Line.append`
- `black.nodes.whitespace`
- `black.brackets.BracketTracker.mark`
- `black.linegen.LineGenerator.visit_default`
- `black.parsing.stringify_ast`
- `black.comments.generate_comments`
- And maybe some other visitor methods on `black.linegen.LineGenerator` (will need to do analysis on which ones are hot)

**Comment by ichard26** [comment]
OK, another update (that is honestly going to be read by only me for the foreseeable future :slightly_smiling_face:).

I got lazy and didn't want to invest the hours it would take to learn the logic under src/black to shift the logic around for performance wins (as like I did for src/blib2to3) with the same confidence of not breaking things. So I only attempted optimizations I knew would help and weren't dangerous. Turns out that was more of a valid strategy than I would have initially assumed since it netted me a ~6% additional perf win :tada: (compared to interpreted main, NOT compiled optimization two). Fmt-fast is now 2.14x faster which is just amazing.

Anyway, other than getting distracted by diff-shades, I looked into the gcc issue to find a potential workaround. I found one, but it was useless :slightly_smiling_face: Collecting numbers for both gcc-10 and clang showed that gcc fails in basically all departments:
- takes almost twice as long to compile Black AND definitely uses 200%+ more memory
- had no meaningful difference in generated binary size all while being far more picky about the C code it's given
- oh and of course produced binaries that were around 8% slower :upside_down_face: 

I'm honestly happy that GCC failed to compile on that parser setup code, clang (for Linux) is so much better.

Time for blackbench related work + benchmarking + stability checks.
